### ISSUE TITLE
SELinux context switching

### ISSUE THREAD

**Issue by henrik-lindqvist** [topjohnwu/Magisk#607]
Please provide a way choose the SELinux context of the provided shell.
SuperSU has an --context <context> switch for its su binary.

**Comment by topjohnwu** [comment]
Why do you need this? Switching contexts is pointless at this point, it is a weird hack CF introduced back in the days and I really don't get the point why he did so. The `magisk` context should already be in "GOD Mode". If there is any additional rules you need, report it here and if it is appropriate then I'll add it.

**Comment by RikkaW** [comment]
As a complement to topjohnwu's words, if you really have special requirements to run your executable in a different context, use `dlopen dlsym` to call `setcon` in `libselinux` in your executable to switch context is ok.

**Comment by henrik-lindqvist** [comment]
@topjohnwu My app use "exec app_process ..." to launch an binder "service" from the su shell. It doesn't work with Magisk unless the users change to Permissive, i suspect it's caused by a too restrictive context.

@RikkaW I'll likely try that, but my app would have to rely on NDK, maybe the runcon command will suffice.  However, from what i understand, "transitioning" to another context has a lot prerequisites which an "app_process" likely doesn't fulfill?

**Comment by DenyDarko** [comment]
I'm pretty sure the denial log from SE will be extremely useful

**Comment by RikkaW** [comment]
@henrik-lindqvist This is simple, you just need use `supolicy` tool provided by magisk to patch SELinux rules.

`supolicy --live "allow * magisk binder *"` 
`supolicy --live "allow * su binder *"` (for older magisk)
`supolicy --live "allow * default_android_service service_manager find"` (if you add your service into ServiceManager and app will use `ServiceManager.getService` to find it)
(All the first * can be replace by `untrusted_app` and others depends on target api of apps which need to access that service)
(The last * can be replaced by `call`, not sure)

Btw, `logcat | grep 'avc: '` is extremely useful

**Comment by henrik-lindqvist** [comment]
@RikkaW I don't like the idea of "hacking", making (permanent) global changes, to user devices just to get the root features in my app working. Why would changing policies be necessary if the Magisk su binary is in "GOD Mode" as @topjohnwu say? Maybe it's just lacking "binder"?

**Comment by topjohnwu** [comment]
@henrik-lindqvist that's what I'm saying, run in permissive mode, and give me the avc audit denial logs so I can add the rules you needed directly into Magisk :)

**Comment by henrik-lindqvist** [comment]
@topjohnwu Here's the denial which seem to prevent it from working:
09-17 20:14:59.193  7147  7147 I amalab.automate: type=1400 audit(0.0:156): avc: denied { call } for scontext=u:r:untrusted_app_25:s0:c512,c768 tcontext=u:r:magisk:s0 tclass=binder permissive=1

Here's another denial, raised when passing a binder object:
09-17 20:15:16.320  7147  7147 I amalab.automate: type=1400 audit(0.0:157): avc: denied { transfer } for scontext=u:r:untrusted_app_25:s0:c512,c768 tcontext=u:r:magisk:s0 tclass=binder permissive=1

My app use a lot of root features so there may be more, i'll let you know, but they're less important.
If i wanted to inform user on how to make it work with their current Magisk install, without changing to Permissive, how would such commands look, as @RikkaW suggested?
    supolicy --live "allow * magisk binder call"
    supolicy --live "allow * magisk binder transfer"
    supolicy --live "allow * su binder call"
    supolicy --live "allow * su binder transfer"

**Comment by topjohnwu** [comment]
@henrik-lindqvist rules can be grouped together:
```
magiskpolicy --live "allow * magisk binder call" "allow * magisk binder transfer"
```
Use `magiskpolicy` to specifically use Magisk's own tool. `supolicy` is an alias to SuperSU's tool, unless you target to support SuperSU as well.
Also, Magisk has been swtiching to using `magisk` for quite some while now, I don't think adding `su` to the mix has much use.

**Comment by henrik-lindqvist** [comment]
@topjohnwu Thanks. Do you have any insight on why the "binder" permission-class was missing/excluded?

**Comment by topjohnwu** [comment]
@henrik-lindqvist I only make the `magisk` context permissive (so, any transaction from `magisk` is allowed). Some other rules are added for magisk to function properly. I'm not an advanced root developer so i don't know all rules that are required in the wild.

The live patch does not survive a reboot. It loads the patched sepolicy at runtime.

**Comment by henrik-lindqvist** [comment]
@topjohnwu With the missing "binder" policies added, consider this issue fixed. I'll let you know, in new tickets, if i find any more policies to add. Any estimate on which upcoming version that will include the fix, 17.2?

Hmm, live patching not persisting reboot will make it difficult for users to apply/manage themselves. Oh well i'll figure something out. Thanks again.
